src/core/cpuid.d

```diff
@@ -651,130 +676,157 @@ void getCpuInfo0B()
+        } else {
+            uint ext;
+            asm pure nothrow @nogc {
+                mov EAX, 7; // Structured extended feature leaf.
+                mov ECX, 0; // Main leaf.
+                cpuid;
+                mov ext, EBX; // HLE, AVX2, RTM, etc.
+            }
+            cf.extfeatures = ext;
         }
-
-        cf.extfeatures = ext;
     }
 
     if (cf.miscfeatures & OSXSAVE_BIT)
     {
-        asm pure nothrow @nogc {
+        version(GNU) asm pure nothrow @nogc {
+            "xgetbv" : "=a" a, "=d" d : "c" 0;
+        } else asm pure nothrow @nogc {
             mov ECX, 0;
             xgetbv;
             mov d, EDX;
             mov a, EAX;
         }
         cf.xfeatures = cast(ulong)d << 32 | a;
     }
+
     cf.amdfeatures = 0;
     cf.amdmiscfeatures = 0;
     if (max_extended_cpuid >= 0x8000_0001) {
-        asm pure nothrow @nogc {
-            mov EAX, 0x8000_0001;
-            cpuid;
-            mov c, ECX;
-            mov d, EDX;
+        version(GNU) asm pure nothrow @nogc {
+            "cpuid" : "=a" a, "=c" cf.amdmiscfeatures, "=d" cf.amdfeatures : "a" 0x8000_0001 : "ebx";
+        } else {
+            asm pure nothrow @nogc {
+                mov EAX, 0x8000_0001;
+                cpuid;
+                mov c, ECX;
+                mov d, EDX;
+            }
+            cf.amdmiscfeatures = c;
+            cf.amdfeatures = d;
         }
-        cf.amdmiscfeatures = c;
-        cf.amdfeatures = d;
     }
     // Try to detect fraudulent vendorIDs
     if (amd3dnow) cf.probablyIntel = false;
 
-    stepping = a & 0xF;
-    uint fbase = (a >> 8) & 0xF;
-    uint mbase = (a >> 4) & 0xF;
-    family = ((fbase == 0xF) || (fbase == 0)) ? fbase + (a >> 20) & 0xFF : fbase;
-    model = ((fbase == 0xF) || (fbase == 6 && cf.probablyIntel) ) ?
-         mbase + ((a >> 12) & 0xF0) : mbase;
-
     if (!cf.probablyIntel && max_extended_cpuid >= 0x8000_0008) {
-        // determine max number of cores for AMD
-        asm pure nothrow @nogc {
-            mov EAX, 0x8000_0008;
-            cpuid;
-            mov c, ECX;
-        }
         //http://support.amd.com/TechDocs/25481.pdf pg.36
         cf.maxCores = 1;
         if (hyperThreadingBit) {
+            // determine max number of cores for AMD
+            version(GNU) asm pure nothrow @nogc {
+                "cpuid" : "=a" a, "=c" c : "a" 0x8000_0008 : "ebx", "edx";
+            } else asm pure nothrow @nogc {
+                mov EAX, 0x8000_0008;
+                cpuid;
+                mov c, ECX;
+            }
             cf.maxCores += c & 0xFF;
         }
     }
 
     if (max_extended_cpuid >= 0x8000_0004) {
-        char *procptr = cf.processorNameBuffer.ptr;
-        version(D_InlineAsm_X86)
+        uint* pnb = cast(uint*)cf.processorNameBuffer.ptr;
+        version(GNU)
+        {
+            asm pure nothrow @nogc { "cpuid" : "=a" pnb[0], "=b" pnb[1], "=c" pnb[ 2], "=d" pnb[ 3] : "a" 0x8000_0002; }
```

Comment:
I find this style useful here, where output and input are obvious and the pattern is easy on the eyes. What do others think ?